metrics/opentelemetry.go

```diff
@@ -0,0 +1,226 @@
+		exporter,
+		sdkmetric.WithInterval(15*time.Second),
+	)
+
+	provider := sdkmetric.NewMeterProvider(
+		sdkmetric.WithReader(reader),
+		sdkmetric.WithResource(res),
+	)
+	otel.SetMeterProvider(provider)
+
+	return func() {
+		shutdownCtx, cancel := context.WithTimeout(context.Background(), gracePeriod)
+		defer cancel()
+		if err := provider.Shutdown(shutdownCtx); err != nil {
+			logger.Error("otlp metric-provider failed to shutdown", "err", err)
+		}
+	}, nil
+}
+
+func initGlobalTracer(ctx context.Context, res *resource.Resource, cfg config.Config, logger *log.Logrus) (func(), error) {
+	exporter, err := otlptrace.New(ctx, otlptracegrpc.NewClient(
+		otlptracegrpc.WithEndpoint(cfg.OtelCollectorAddr),
+		otlptracegrpc.WithInsecure(),
+		otlptracegrpc.WithCompressor(gzip.Name),
+	))
+	if err != nil {
+		return nil, fmt.Errorf("create trace exporter: %w", err)
+	}
+
+	tracerProvider := sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(consistent.ProbabilityBased(cfg.OtelTraceSampleProbability)),
+		sdktrace.WithResource(res),
+		sdktrace.WithSpanProcessor(sdktrace.NewBatchSpanProcessor(exporter)),
+	)
+
+	otel.SetTracerProvider(tracerProvider)
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
+		propagation.TraceContext{},
+		propagation.Baggage{},
+	))
+
+	return func() {
+		shutdownCtx, cancel := context.WithTimeout(context.Background(), gracePeriod)
+		defer cancel()
+		if err := tracerProvider.Shutdown(shutdownCtx); err != nil {
+			logger.Error("otlp trace-provider failed to shutdown", "err", err)
+		}
+	}, nil
+}
+
+func getSliceStringPluginNames(prs []recipe.PluginRecipe) []string {
+	var res []string
+	for _, pr := range prs {
+		res = append(res, pr.Name)
+	}
+
+	return res
+}
+
+// RecordRun records a run behavior
+func (m *OtelMonitor) RecordRun(ctx context.Context, run agent.Run) {
+	span := trace.SpanFromContext(ctx)
+	span.SetAttributes(attribute.String("recipe_name", run.Recipe.Name))
+
+	m.recipeDuration.Record(ctx,
+		int64(run.DurationInMs),
+		metric.WithAttributes(
+			attribute.String("extractor", run.Recipe.Source.Name),
+			attribute.StringSlice("processors", getSliceStringPluginNames(run.Recipe.Processors)),
+			attribute.StringSlice("sinks", getSliceStringPluginNames(run.Recipe.Sinks)),
+			attribute.Bool("success", run.Success),
+		))
+
+	m.extractorRetries.Add(ctx, int64(run.ExtractorRetries))
+
+	m.assetsExtracted.Add(ctx,
+		int64(run.AssetsExtracted),
+		metric.WithAttributes(
+			attribute.String("extractor", run.Recipe.Source.Name),
+			attribute.StringSlice("processors", getSliceStringPluginNames(run.Recipe.Processors)),
+			attribute.StringSlice("sinks", getSliceStringPluginNames(run.Recipe.Sinks)),
+		))
+}
+
+// RecordPlugin records a individual plugin behavior in a run, this is being handled in otelmw
+func (*OtelMonitor) RecordPlugin(_ context.Context, _ agent.PluginInfo) {}
+
+func (m *OtelMonitor) RecordPluginRetryCount(ctx context.Context, pluginInfo agent.PluginInfo) {
```

Comment:
Should we rename this to `RecordSinkRetryCount`?

Reply:
we can. currently OtelMonitor are being used by plugins regardless of plugin types so thinking of this, updated.